docs/cluster_api.md

```diff
@@ -0,0 +1,95 @@
+# Cluster Operator API
+
+## Overview
+
+The cluster operator API defines a set of resources that allow the instantiation of OpenShift clusters and subsequent
+updates to those clusters based on changes to said resources.
+
+The general approach for these resources is to follow the Kubernetes model of a spec section that is end-user controlled and
+a status section that represents the current state of the resources. Changes to the spec will result in the necessary actions
+by the system to reconcile the current state of the system with what the user has specified.
+
+## Resources
+
+### Cluster
+
+The cluster resource should allow 2 levels/scopes of spec:
+- Cluster-wide configuration
+- NodeGroup configuration
+
+Furthermore, each scope can be described in 2 parts:
+- Hardware spec (needed to instantiate the proper virtual hardware)
+- OpenShift config (openshift-specific configuration)
+
+For the Status part, if the cluster has been instantiated, we need to know how to access it (Kubeconfig may be sufficient)
+
+In more detail:
+
+#### Spec:
+* Cluster-wide configuration:
+  * ObjectMetadata (inline ObjectMetadata)
+  * TypeMetadata (inline TypeMetadata)
+  * Hardware-spec
+    * Type (AWS, Azure, GCE, etc) [string]
```

Comment:
We already have an "API" as defined by the variables that openshift-ansible accepts.  You will then have to transform this API into that API before openshift-ansible will run.  Why not define your API *exactly* as the the openshift-ansible variables would be, and formalize those more strongly?  

Articulate how having these spec variables is going to solve a user problem (you have no justification here for why), vs creating more work and maintenance.  Are any of these values going to be changeable post create?  Which ones?  We put things in API fields so that end users can set them.  Which ones of these will be set?

Is this API going to be the only source of truth?  If not, that's a third API you will be converting from and to (which is probably at least 1 too many).  If it is the source of truth, do you expect to support being able to create and recreate this entity and deal with how changes would prevent it from running?

Reply:
> We already have an "API" as defined by the variables that openshift-ansible accepts. You will then have to transform this API into that API before openshift-ansible will run. Why not define your API exactly as the the openshift-ansible variables would be, and formalize those more strongly?

We need to have separate identifiable entities that we can work with (a cluster, machine set, etc). The ansible API is a flat list of variables, not exactly sure what you mean by having them match exactly. I agree we should avoid inventing new things so things are harder to maintain, and I don't think there's anything wildly different here than what you can specify in ansible.

Reply:
> Articulate how having these spec variables is going to solve a user problem (you have no justification here for why), vs creating more work and maintenance. Are any of these values going to be changeable post create? Which ones? We put things in API fields so that end users can set them. Which ones of these will be set?

I can clarify a little bit more, but the basic idea is that the spec portion is end-user changeable and to "make it so" we will do what's necessary, including spinning down your cluster and bring a new one up.

Reply:
> Is this API going to be the only source of truth?

For clusters managed by this API, yes, that's the intent.

Reply:
I would pretty strongly disagree that the openshift-ansible variable set should be the 'API'. We are likely going to need to merge this and the upstream cluster api. So while we need to recognize what openshift-ansible is asking for we need to also recognize that in the long run we will likely need to do translations. So making this API seem as sane as possible will be the most useful as we try to reconcile with upstream.

Reply:
I'm not saying it should be "the" API.  I'm saying it is our existing provision API.  It already has to be stable for customers.  Because it must be stable for customers, it should also be stable for other consumers like cluster-operator.  Cluster operator is mapping part of the provision API up a level.  Like any "abstraction" API there will be constant urge to expose things up from the lower level API (i.e. node tuning parameters), and the more delta there is between how the ansible API is designed and how the cluster API is designed there will be more bugs, confusion, documentation, and human errors, which result in tech debt, failures in customer environments, and unhappy users.

So while you may have to align with upstream, you also need to reduce the amount of conceptual delta between ansible and the operator, force ansible to declare and support stable APIs for you, and test those APIs.  Even the naming and structure is important - what unit does ansible accept, how about nils, what happens if the value is unrecognized.  It's insufficient to design this API without understanding the API we expose today, even if we wish to change it.

Reply:
Re: layers of abstraction, with cluster operator here are the layers:

1. cluster api (source of truth)
2. ansible via install and update
3. master-config, node-config
4. kube component flags and config

Customers will continue to use #2, and we are going to work at and evolve collapsing 3-4 (although our approach may change).  It's really important that when you have to expose something up from 3-4 through 2 to 1, that the mapping be as simple as possible.

Reply:
> I can clarify a little bit more, but the basic idea is that the spec portion is end-user changeable and to "make it so" we will do what's necessary, including spinning down your cluster and bring a new one up.

Talked with cesar in person, but it's extremely unlikely any sort of core cloud configuration can be changed.  None of your IP addresses are going to be able to move, for one thing, and data PV transfer is far more complex.  Probably better to just say - certain fields that are fundamental are not changeable by default, and you identify the fields that *can* change with good use cases.  I don't care what the use cases are (well, I do), but make sure you are justifying *why* they can change.